rtbkit/common/augmentor_interface.cc

```diff
@@ -0,0 +1,86 @@
+/* augmentor_interface.cc
+   Mathieu Stefani, 20 janvier 2016
+   Copyright (c) 2015 Datacratic.  All rights reserved.
+   
+*/
+
+#include "augmentor_interface.h"
+
+using namespace Datacratic;
+
+namespace RTBKIT {
+
+AugmentorInterface::AugmentorInterface(
+    ServiceBase& parent, const std::string& serviceName)
+  : ServiceBase(serviceName, parent),
+    inbox(65536)
+{ }
+
+AugmentorInterface::AugmentorInterface(
+    std::shared_ptr<ServiceProxies> proxies,
+    const std::string& serviceName)
+  : ServiceBase(serviceName, std::move(proxies)),
+    inbox(65536)
+{ }
+
+void AugmentorInterface::init(){
+    inbox.onEvent = [&] (AugmentMessage&& message)
+        {
+            doSendAugmentMessage(
+                message.instance,
+                std::move(message.augmentorName),
+                message.auction,
+                message.agents,
+                std::move(message.date));
+        };
+
+    addSource("AugmentorInterface::inbox", inbox);
+}
+
+void AugmentorInterface::sendAugmentMessage(
+            const std::shared_ptr<AugmentorInstanceInfo>& instance,
+            const std::string& augmentorName,
+            const std::shared_ptr<Auction>& auction,
+            const std::set<std::string>& agents,
+            Datacratic::Date date){
+
+    AugmentMessage m;
+    m.instance = instance;
+    m.augmentorName = std::move(augmentorName);
```

Comment:
Because it has no reason to complain. Say that you have two overloads:

```cpp
std::string(const std::string& other);
```

and
```cpp
std::string(std::string&& other)
```

Now if you write:

```cpp
const std::string foo("foo");
std::string bar(std::move(foo));
```

During overload resolution, the compiler will first look at all the potential matches.
`std::move(foo)` is just an rvalue-cast, since `foo` is const, you will get `const std::string&&`.
Now, rules say that non-const rvalues only bind to non-const rvalue references and const-lvalue references, which invalidates the second overload, thus the compiler will select the first one.